docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+   which prevents the Raft database encryption key
+   from being stored plaintext on disk.
+
+   * Kubernetes:
+   Kubernetes is not officially supported by the EdgeX project.
+   Kubernetes also supports the secrets volume approach,
+   though the secrets volume can be mounted anywhere in the container namespace.
+   For an example of Kubernetes secrets volumes, see the
+   [Kubernetes secrets documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+   Secrets distributed in this manner become part of the `etcd` database,
+   and thus it becomes necessary to specify a
+   [KMS provider for data encryption](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/)
+   to prevent `etcd` from storing plaintext versions of secrets.
+
+   * Docker:
+   Use of a RAM-based temporary file system is desirable to protect secrets,
+   but only Linux platforms support `tmpfs` volumes for secret sharing:
+   generally, `/run` on a Linux system is a `tmpfs` volume.
+   For reasons of portability, `/tmp/edgex/secrets` has been chosen as a
+   cross-platform alternative to `/run`:
+   on the MacOS Docker implementation,
+   `/tmp` is a white-listed shared host directory,
+   while `/run` is not.
+   For production hardening, `/tmp` should be replaced by `/run`
+   where it appears in `docker-compose` files.
+   If the container runtime in use does not support host fileystem bind mounts,
+   then a separate Docker secrets volume should be created for each service.
+
+   * Snaps:
+   Note that due to an existing [bug](https://bugs.launchpad.net/snap-confine/+bug/1620442),
+   `$XDG_RUNTIME_DIR` is set to a snap-specific path
+   (`/run/user/1000/snap.`_$SNAP_NAME_`/`),
+   but the directory is **neither created nor accessible to the snap**.
+   Because it may be necessary to share several secrets outside of the snap
+   --such as secret store tokens--
+   it is desirable to place them in a location that can be mapped to
+   a snap content interface, such as `$SNAP_DATA/secrets`.
+   Although not a `tmpfs`, this is the best available option at this time.
+   The snap bug should be escalated in priority.
+
+## Consequences
+
+As the existing implementation is not fully-compliant with this ADR,
+significant scope will be added to current and future EdgeX releases
+in order to bring the project into compliance.
+
+List of needed improvements:
+
+- PKI private keys
+  * All: Move to using Vault as system of origin for the PKI instead of the standalone `security-secrets-setup` utility.
+  * All: Cache the PKI for Consul and Vault on persistent disk; rotate occasionally.
+  * All: Investigate hardware protection of cached Consul and Vault PKI secret keys.
+  * Snaps: Relocate PKI into the secrets volume instead of its own folder.
+
+- Secret store master password
+  * All: Enable hooks for hardware protection of secret store master password.
+
+- Secret store per-service authentication tokens
```

Comment:
For snaps, we need to introduce a content interface that will allow application and device service snaps to read secret-store tokens from the main edgexfoundry snap.

Reply:
> For snaps, we need to introduce a content interface that will allow application and device service snaps to read secret-store tokens from the main edgexfoundry snap.

Agree.  This is another usage of the **to-be-named location**.

Reply:
_SECRETSLOC_